---
title: "Team 5 Group Project"
author: "Mary Weinrich"
date: "11/5/19"
output:
  html_document:
    df_print: paged
  pdf_document: default
  word_document: default
editor_options:
  chunk_output_type: console
---

Notes: 
  - You do not have to put all of your team members' code into a single file. I have included all 5 analyses just for your information. You only need the code for your analysis.
  - The tasks include both coding and written interpretation. 
  - Please knit to word document -- it will make it easier to combine your results with your team members in to the single manuscript (submitted in GP4).

## Setup

### Load packages

Add whatever additional packages you need for your analysis

```{r setup, include=FALSE}
### EDIT!!!

### We use the code chunk option "include=FALSE" because we don't need to print this information

### Global knitr options
knitr::opts_chunk$set(echo = TRUE)

### Load packages/libraries that we will need
library(tidyverse)
library(viridis)       # Better plot colors
library(ggbiplot)      # PCA plot
library(corrr)         # Correlations
library(ggridges)      # Density ridge plot
library(ggpubr)        # Fancy plots
library(psych)         # Distributions
#  edit! Add whatever additional packages you need here (if you haven't loaded them, RMarkdown should alert you when you go to "knit" the RMarkdown to a report)
```


### Custom ggplot theme

So that we don't need to add this code to all ggplots individually. Feel free to use or not use, and to modify however you wish.

```{r theme}
### DON'T EDIT CODE IN THIS CHUNK

theme_custom <- theme_bw() +
  
  # if we have a plot title or subtitle, let's center it
  theme (
    plot.title = element_text(hjust = 0.5),
    plot.subtitle = element_text(hjust = 0.5)
  ) 
theme_set(theme_custom)

### We'll make the viridis color scale our default plotting color palette
scale_colour_continuous <- function(...) {
  scale_colour_viridis_c(...)
}
scale_fill_continuous <- function(...) {
  scale_fill_viridis_c(...)
}
scale_colour_discrete <- function(...) {
  scale_colour_viridis_d(..., begin = 0, end = 0.9)
}
scale_fill_discrete <- function(...) {
  scale_fill_viridis_d(..., begin = 0, end = 0.9)
}
```


### Setwd fix (if needed)

If you are having trouble loading the exprs_tidy file below, manually override the working directory. To do this
  1. In the menu bar, click: Session > Set Working Directory > To Source File Location
  2. Copy the line of code in the console, and paste it into the code chunk below
  
```{r fix_setwd}
### EDIT if necessary

```


### FYI: how I got the data

```{r get_data, eval=FALSE}
### Get list of available datasets
### https://www.bioconductor.org/packages/3.3/bioc/vignettes/TCGAbiolinks/inst/doc/tcgaBiolinks.html#harmonized-data-1
View(getGDCprojects())

### Datasets to use for group project (I picked the ones with smallest sample size and no sex bias)
projects <- c(
  "TCGA-ACC",
  "TCGA-CHOL", 
  "TCGA-DLBC", 
  "TCGA-KICH", 
  "TCGA-MESO", 
  "TCGA-UVM"
)

phenoList <-  vector(mode = "list", length = length(projects))
names(phenoList) <- projects
exprsList <-  vector(mode = "list", length = length(projects))
names(exprsList) <- projects
for (i in projects) { 
  ### Get data (in summarized experiment ["se"]  format)
  query <- GDCquery(
    project = i, 
    data.category = "Transcriptome Profiling", 
    data.type = "Gene Expression Quantification", 
    workflow.type = "HTSeq - FPKM"
  )
  GDCdownload(query)
  se <- GDCprepare(query)
  
  ### Extract phenoData and remove columns that either are all different or all consistent
  pheno_full <- as.data.frame(colData(se))
  pheno <- janitor::remove_constant(pheno_full)
  
  ### Extract exprs matrix and remove lowly expressed
  exprs_full <- assay(se)
  keep <- rowSums(exprs_full > 1) >= 10
  exprs <- exprs_full[keep, ]

  ### Shorten the sample id
  rownames(pheno) <- abbreviate(gsub("TCGA-OR-", "", rownames(pheno)), method = "both")
  pheno$id <- rownames(pheno)
  colnames(exprs) <- abbreviate(gsub("TCGA-OR-", "", colnames(exprs)), method = "both")
  
  ### Remove extra columns (not groups)
  pheno$sample <- pheno$id
  pheno$id <- NULL
  remove_cols <- c(
    "patient", "updated_datetime", "updated_datetime.x", "updated_datetime.y", 
    "barcode", "diagnosis_id", "demographic_id", "exposure_id", "bcr_patient_barcode", 
    "morphology", "treatments", 
    "days_to_birth", "days_to_last_follow_up", "days_to_death",
    "year_of_birth", "year_of_diagnosis", "year_of_death"
  )
  pheno <- pheno[ , !(colnames(pheno) %in% remove_cols)]
  pheno <- pheno[ , !(colnames(pheno) %in% colnames(pheno)[grep("_CHOL_del|_CHOL_amp|subtype_", colnames(pheno))])]

  ### Save
  saveRDS(exprs, paste0(i, "_exprs.rds"))
  saveRDS(pheno, paste0(i, "_pheno.rds"))
  
  ### Add to list
  exprsList[[i]]  <- exprs
  phenoList[[i]] <- pheno
  
  ### Clean up
  rm(exprs)
  rm(exprs_full)
  rm(pheno)
  rm(pheno_full)
  rm(keep)
}

### Save
saveRDS(exprsList, "all_exprs.rds")
saveRDS(phenoList, "all_pheno.rds")

### Look at
sapply(exprsList, dim)
sapply(phenoList, dim)
sapply(phenoList, names)

### Write out names
rbind(
  paste("ACC:", toString(sort(names(phenoList$`TCGA-ACC`)))),
  paste("CHOL:", toString(sort(names(phenoList$`TCGA-CHOL`)))),
  paste("DLBC:", toString(sort(names(phenoList$`TCGA-DLBC`)))),
  paste("KICH:", toString(sort(names(phenoList$`TCGA-KICH`)))),
  paste("MESO:", toString(sort(names(phenoList$`TCGA-MESO`)))),
  paste("UVM:", toString(sort(names(phenoList$`TCGA-UVM`))))
) %>%
  writeLines("sample_variables.txt")
```


## [EDIT AS TEAM] Pre-process data 

Your entire team should use the same code for this section!

### Load your dataset [edit!]

```{r load_data}
### EDIT: You need to insert your dataset file names in the quotes below

exprs <- readRDS(
  "TCGA-MESO_exprs.rds"     # EDIT: insert your *_exprs.rds dataset's file name here
)
pheno <- readRDS(
  "TCGA-MESO_pheno.rds"     # EDIT: insert your *_pheno.rds dataset's file name here
)
```

### Pick your group (variable of interest) [edit!]

This should be a variable that is categorical with at least 2 categories and at least 3 samples in each category Use colnames(pheno) to find out what variable options you have. You can use one of the descriptive summary functions (from AE3) to determine how many categories there are for each group, and how many samples there are for each category.

```{r select_group}
### EDIT!! Copy your variable of interest into a new column called "group". This will help generalize/simplify your project's code

pheno$group <- pheno$gender  # EDIT: insert your variable's column name here
```

### Convert expression matrix to tidy 

```{r tidy}
### Don't edit

### Create tibble with expression and pheno data
tidy <- exprs %>% 
  
  # Convert expression matrix to tidy tibble
  as_tibble(rownames = "gene") %>%
  gather("sample", "fpkm", -gene)  %>%

  # Add phenoData
  inner_join(pheno, by = "sample")
```

### Filter samples [edit if appropriate]

Check for samples with missing data for your "group"

```{r filter_samples}
### Don't edit

### You can check this using the following (many other ways to check too)
### Make sure no "blanks" either -- 
### sometimes missing data isn't converted to an NA but instead is just blank
summary(as.factor(tidy$group)) # look for blanks, no data, etc. categories
table(is.na(tidy$group))
```

Remove samples with missing data (or no data; i.e. "not reported") for your "group"

```{r filter_samples}
### EDIT (if you have any samples with NA or blank for your group)

tidy <- tidy %>% 
  filter(
    
    # EDIT: add filter(s) here to remove NAs and samples with no data, below are common examples of how missing data is listed in phenoData. Make sure you don't need any additional filters to remove missing data for your "group"
    
    group != "not reported",    # some samples may have missing data listed as "not reported"
    !is.na(group),              # some may have missing data listed as NA
    group != ""                 # some may have blanks
    
  )
#Our group wants to focus on the role of the BAP1 gene so tidy will be filtered for this
BAP1 <- tidy %>%
  filter(gene == "ENSG00000163930")
```

### Convert back to matrix (some analyses use matrices rather than tidy data)

```{r matrices}
### Don't edit

### Convert expression data to wide/matrix
exprs_mat <- tidy %>%
  select(gene, sample, fpkm) %>%
  spread(sample, fpkm, fill = 0) %>%
  column_to_rownames("gene") %>%
  as.matrix()

### Convert phenoData to wide table
pheno_mat <- tidy %>%
  select(-one_of(c("gene", "fpkm"))) %>%
  distinct() %>%
  data.frame()
rownames(pheno_mat) <- pheno_mat$sample

### Fix order of samples, so they match
table(colnames(exprs_mat) == rownames(pheno_mat)) # check
pheno_mat <- pheno_mat[order(rownames(pheno_mat), colnames(exprs_mat)), ] # fix
table(colnames(exprs_mat) == rownames(pheno_mat)) # check
```

### Filter genes

Here, we'll only keep genes that are expressed at >= 1 FPKM in at least 20 samples.

```{r filter_genes_fixed}
### Don't edit

### Count # genes pre-filtering
nrow(exprs_mat)

exprs_mat <- exprs_mat[rowSums(exprs_mat > 1) > 20, ]

### Count # genes post-filtering
nrow(exprs_mat)
```


## [one team member does this] Team Member #3: Hierarchical clustering

Tasks:
  a. Determine ideal clustering methods (including # clusters)
  b. Generate a final figure of the clustering 
  c. Provide table summarizing the number of samples in each cluster and the breakdown of samples by your group of interest
  d. Interpret the clustering results

[enter code here, see AE3 to get started!]

```{r}
#additional packages to turn on
library(ggdendro)
library(ggrepel)
library(factoextra)
library(cluster)
library(dendextend)
```


```{r}
#Find optimal number of clusters = 2

fviz_nbclust(BAP1, FUN = hcut, method = "wss")
```

```{r}
#Establish the number of clusters
n_clust <- 2

```

```{r}
#more attempts at dendrograms

### Calculate HC
hc3_final <- agnes(BAP1, method = "ward", metric = "euclidean")

### Cut tree
hc3_final_clusters <- cutree(hc3_final, k = n_clust)

```

```{r}
### View tree 
clust.df <- data.frame(
  label = names(hc3_final_clusters), 
  cluster = factor(hc3_final_clusters)
)
dendr <- as.ggdend(as.dendrogram(hc3_final))
dendr[["labels"]] <- merge(
  dendr[["labels"]], 
  clust.df, 
  by = "label"
)
ggplot() + 
  geom_segment(
    data = segment(dendr), 
    aes(x = x, y = y, xend = xend, yend = yend)
  ) + 
  
  # Label clusters
  geom_text(
    data = label(dendr), 
    aes(x, y, label = label, color=cluster), 
    size = 3, hjust = 0
  ) +
  
  # Flip axes
  coord_flip() +

  # Formating 
  scale_y_reverse(expand = c(0.2, 0)) +    
    theme_classic() +   
    theme(     
      axis.line.y = element_blank(),     
      axis.ticks.y = element_blank(),     
      axis.text.y = element_blank(),     
      axis.title = element_blank()
      )
  
```


```{r}
#make heatmap to see if theres a pattern in the expression
heatmap(exprs_mat)
```

